pkg/operations/cordondrainvm.go

```diff
@@ -228,3 +235,154 @@ func (o *drainOperation) deletePods(pods []v1.Pod) error {
+
+			if podAfterDrain == nil || len(podAfterDrain.Status.Conditions) == 0 {
+				logger.Debugf("Pod %s is not scheduled yet.", podName)
+				continue
+			}
+
+			firstCondition := podAfterDrain.Status.Conditions[0]
+			if string(firstCondition.Type) == "PodScheduled" && firstCondition.Reason == "Unschedulable" {
+				logger.Infof("Pod %s cannot not be scheduled: %s Stop monitoring volume attachment.", podName, firstCondition.Message)
+				if match, _ := regexp.MatchString("0/[0-9]+ nodes are available:.*exceed max volume count", firstCondition.Message); match {
+					logger.Info("Nodes exceed max volume count, cancelling the volume attachment monitoring.")
+					cancel()
+				}
+				return
+			}
+
+			nodeName := podAfterDrain.Spec.NodeName
+			node, err := client.GetNode(nodeName)
+			if err != nil {
+				logger.Errorf("Failed to get node '%s': %s", nodeName, err.Error())
+				return
+			}
+
+			done := true
+			provisioner := "kubernetes.io/azure-disk"
+			logger.Debugf("Found new pod %s scheduled on node '%s'", podName, nodeName)
+			for _, volume := range podAfterDrain.Spec.Volumes {
+				pvcSpec := volume.VolumeSource.PersistentVolumeClaim
+				if pvcSpec != nil {
+					pvc, err := client.GetPersistentVolumeClaim(namespace, pvcSpec.ClaimName)
+					if err != nil {
+						logger.Errorf("Failed to get PVC %s in namespace %s: %s", pvcSpec.ClaimName, namespace, err.Error())
+						return
+					}
+
+					pv, err := client.GetPersistentVolume(pvc.Spec.VolumeName)
+					if err != nil {
+						logger.Errorf("Failed to get PV %s: %s", pvc.Spec.VolumeName, err.Error())
+						return
+					}
+
+					logger.Debugf("Pod %s is using PV %s on node '%s'", podName, pv.Name, nodeName)
+					if pv.ObjectMeta.Annotations["pv.kubernetes.io/provisioned-by"] == provisioner {
+						attached := false
+						for _, volumeAttached := range node.Status.VolumesAttached {
+							if string(volumeAttached.Name) == provisioner+"/"+pv.Spec.AzureDisk.DataDiskURI {
+								logger.Debugf("Volume %s is attached to node %s.", pv.Spec.AzureDisk.DataDiskURI, nodeName)
+								attached = true
+								break
+							}
+						}
+
+						if !attached {
+							logger.Debugf("Volume %s is not attached to node %s yet.", pv.Spec.AzureDisk.DataDiskURI, nodeName)
+							done = false
+							break
+						}
+					}
+				}
+			}
+
+			if !done {
+				logger.Debugf("Not all volumes for Pod %s are attached to new node %s.", podName, nodeName)
+				continue
+			} else {
+				logger.Infof("All volumes for Pod %s are attached to new node %s.", podName, nodeName)
+				return
+			}
+		}
+	}
+}
+
+// WaitForDisksAttached waits for disks are re-attached to new nodes and volumes are ready to be used by pods.
+func WaitForDisksAttached(podsForDeletion []v1.Pod, client armhelpers.KubernetesClient, logger *log.Entry, args ...time.Duration) error {
+	if len(podsForDeletion) == 0 {
+		return nil
+	}
+
+	timeout := time.Hour * 1
+	interval := time.Minute * 1
+	if len(args) > 0 {
+		timeout = args[0]
```

Comment:
Looks like timeout is not used in the wait function. What if the VM size is not large enough to accept the managed disk?